hs/src/Reach/AST/SL.hs

```diff
@@ -164,94 +163,120 @@ data SLVal
+    ((ST_Fun f), (ST_Fun f2)) -> equiv f f2
+    ((ST_UDFun f), (ST_UDFun f2)) -> equiv f f2
+    ((ST_Type s1), (ST_Type s2)) -> equiv s1 s2
+    ((ST_Refine _sLType v _m), (ST_Refine _sLType2 v2 _m2)) -> equiv v v2
+    _ -> False
 
 instance Equiv DLVar where
   equiv (DLVar _ sl _dl i) (DLVar _ sl2 _dl2 i2) =
     equiv (fmap snd sl) (fmap snd sl2) && i == i2
 
+instance Equiv DLConstant where
+  equiv DLC_UInt_max DLC_UInt_max = True
+
 instance Equiv SLVal where
-  equiv (SLV_Null _ _) (SLV_Null _ _) = True
-  equiv (SLV_Bool _ b1) (SLV_Bool _ b2) = b1 == b2
-  equiv (SLV_Int _ i1) (SLV_Int _ i2) = i1 == i2
-  equiv (SLV_Bytes _ v1) (SLV_Bytes _ v2) = v1 == v2
-  -- Array types can be ignored
-  equiv (SLV_Array _ _ xs) (SLV_Array _ _ ys) = all (\(x,y) -> equiv x y) $ zip xs ys
-  equiv (SLV_Tuple _ v1) (SLV_Tuple _ v2) = v1 == v2
-  equiv (SLV_Struct _ xs) (SLV_Struct _ ys) = xs == ys
-  equiv (SLV_DLC d1) (SLV_DLC d2) = d1 == d2
-  equiv (SLV_DLVar d1) (SLV_DLVar d2) = d1 == d2
-  equiv (SLV_Connector t1) (SLV_Connector t2) = t1 == t2
-  equiv (SLV_Prim p1) (SLV_Prim p2) = p1 == p2
-  equiv (SLV_Kwd k1) (SLV_Kwd k2)  = k1 == k2
-  equiv (SLV_Deprecated d v) (SLV_Deprecated d2 v2) = d == d2 && v == v2
-  equiv SLV_Anybody SLV_Anybody = True
-
-  equiv (SLV_Participant _ s sl dl) (SLV_Participant _ s2 sl2 dl2) = equiv s s2 && equiv sl sl2 && equiv dl dl2
-  equiv (SLV_RaceParticipant _ slSet1) (SLV_RaceParticipant _ slSet2) = equiv slSet1 slSet2
-
-  equiv (SLV_Map v1) (SLV_Map v2) = equiv v1 v2
-  equiv (SLV_Data _ _ _ _) _ = False
-  equiv (SLV_Form f1) (SLV_Form f2)= equiv f1 f2
-  equiv (SLV_Type t1) (SLV_Type t2) = equiv t1 t2
-  equiv (SLV_Object _ _ slenv1) (SLV_Object _ _ slenv2) = equiv slenv1 slenv2
-
-  -- Closures are the only uninspected structures
-  equiv (SLV_Clo _ _ _) _ = False
-  equiv _ _ = False
+  equiv a b = case (a,b) of
+    ((SLV_Null _ _), (SLV_Null _ _)) -> True
+    ((SLV_Bool _ b1), (SLV_Bool _ b2)) -> b1 == b2
+    ((SLV_Int _ i1), (SLV_Int _ i2)) -> i1 == i2
+    ((SLV_Bytes _ v1), (SLV_Bytes _ v2)) -> v1 == v2
+    -- Array types can be ignored
+    ((SLV_Array _ _ xs), (SLV_Array _ _ ys)) -> all (\(x,y) -> equiv x y) $ zip xs ys
+    ((SLV_Tuple _ v1), (SLV_Tuple _ v2)) -> v1 == v2
+    ((SLV_Struct _ xs), (SLV_Struct _ ys)) -> xs == ys
+    ((SLV_DLC d1), (SLV_DLC d2)) -> equiv d1 d2
+    ((SLV_DLVar d1), (SLV_DLVar d2)) -> equiv d1 d2
+    ((SLV_Connector t1), (SLV_Connector t2)) -> t1 == t2
+    ((SLV_Prim p1), (SLV_Prim p2)) -> p1 == p2
+    ((SLV_Kwd k1), (SLV_Kwd k2))  -> k1 == k2
+    ((SLV_Deprecated d v), (SLV_Deprecated d2 v2)) -> d == d2 && v == v2
+    (SLV_Anybody, SLV_Anybody) -> True
+    ((SLV_Participant _ s sl dl), (SLV_Participant _ s2 sl2 dl2)) -> equiv s s2 && equiv sl sl2 && equiv dl dl2
+    ((SLV_RaceParticipant _ slSet1), (SLV_RaceParticipant _ slSet2)) -> equiv slSet1 slSet2
+    ((SLV_Map v1), (SLV_Map v2)) -> equiv v1 v2
+    ((SLV_Data _ _ _ _), _) -> False
```

Comment:
You should compare data instances

Reply:
`SLV_Data SrcLoc (M.Map SLVar DLType) SLVar SLVal`
Do we need to compare anything other than the `SLVal` here?

Reply:
You should compare the last three elements. They are: the variants, the variant, and then the value. For example, we don't want `Left 1` to match `Left 2`, so we compare the last. We don't want `Left 1` to match `Right 1`, so we compare the second to last. And, we don't want `X.Left 1` to match `Y.Left 1` where `X = Left Int | Right Int` and `Y = Left Int | Right Bool`, because they aren't the same kind of `Left`